src/qilib/utils/graph.py

```diff
@@ -0,0 +1,58 @@
+"""Quantum Inspire library
+
+Copyright 2019 QuTech Delft
+
+qilib is available under the [MIT open-source license](https://opensource.org/licenses/MIT):
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+from typing import Set, Any
+
+
+class Node:
+    def __init__(self, content: Any = None) -> None:
+        self.previous_nodes: Set[Node] = set()
+        self.next_nodes: Set[Node] = set()
+        self.content: Any = content
+
+
+class Graph:
+    def __init__(self) -> None:
+        self.start_nodes: Set[Node] = set()
+        self.end_nodes: Set[Node] = set()
+
+    def add_node(self, node: Node) -> None:
+        self._fix_links(node)
```

Comment:
I guess add_node() could implement the code in _fix_links() and then the add_link would just call it instead.

Reply:
I can imagine a situation (e.g. when a possibility to remove links would be added) where you still want to be able to ensure correct linkage. Calling `add_node` in such a situation would be confusing.
Also, I think that `add_node` will get additional functionality in the near future -- possibly to force that the previous/next nodes are empty. Otherwise you could build up a whole structure outside of the graph class and then add just one node of that structure to the graph. That could easily lead to dangling graph branches, where you could reach them from the graph's start_nodes but not from the end_nodes, or vice versa.

Reply:
Yeah, I was thinking about it as well when re-reading the code. which is why I marked it as resolved. I approved the PR.